### PR TITLE
PP-8: Take card details and authorise charge

### PR DESCRIPTION
**Pull request by chaphilippe**
- validation of card details (format and length).
- returning 204 upon success, 400 or 404 upon error.
- the charge status is updated to AUTHORIZATION SUCCESS upon completion of the request.
- added DB migration to run.sh
- DatabaseTestHelper moved to test package

for all details see: https://payments-platform.atlassian.net/browse/PP-8

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/resources/CardDetailsResource.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            return responseWithInvalidCardDetails();
+        }
+
+        chargeDao.updateStatus(chargeId, ChargeStatus.AUTHORIZATION_SUBMITTED);
+
+        //here comes the code for authorization - always successful for the scope of this story.
+
+        chargeDao.updateStatus(chargeId, ChargeStatus.AUTHORIZATION_SUCCESS);
+
+        return Response.noContent().build();
```
Comment (mikemey) [inline src/main/java/uk/gov/pay/connector/resources/CardDetailsResource.java:52]: ~~Should probably return 201 with the new content - see https://github.com/paypal/api-standards/blob/master/api-style-guide.md#create-new-resource (which also applies to sub-resources)~~

will be implemented later.

File: src/main/java/uk/gov/pay/connector/resources/CardDetailsValidator.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package uk.gov.pay.connector.resources;
+
+import java.util.Map;
+
+public class CardDetailsValidator {
+
+    public static boolean isValidCardDetails(Map<String, Object> cardDetails) {
+        return cardDetails.containsKey("cvc") && is3Digits(cardDetails.get("cvc")) &&
+                cardDetails.containsKey("card_number") && is16Digits(cardDetails.get("card_number")) &&
```
Comment (mikemey) [inline src/main/java/uk/gov/pay/connector/resources/CardDetailsValidator.java:9]: The user story description describes the card number with 16 digits and CVC with 3 digits, but there are other valid formats as well:
- card numbers can range between 12-19 digits - see https://en.wikipedia.org/wiki/Bank_card_number#Issuer_identification_number_.28IIN.29 
- CVC numbers can have 3 or 4 digits.

Are we planning on changing this later?

File: src/test/java/uk/gov/pay/connector/it/CardDetailsResourceITest.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    public void shouldAuthoriseChargeForValidCardDetails() throws Exception {
+        String uniqueChargeId = "983794837598475";
+        setupCharge(uniqueChargeId);
+
+        given().port(app.getLocalPort())
+                .contentType(JSON)
+                .body(validCardDetails)
+                .post(cardUrlForChargeId(uniqueChargeId))
+                .then()
+                .statusCode(204);
```
Comment (mikemey) [inline src/test/java/uk/gov/pay/connector/it/CardDetailsResourceITest.java:41]: ~~Response should contain the charge with the card details (masked) - please add assertions here.~~

will be implemented later.

File: src/main/java/uk/gov/pay/connector/resources/CardDetailsValidator.java (hunk 2 above)
Comment (chaphilippe) [inline src/main/java/uk/gov/pay/connector/resources/CardDetailsValidator.java:9]: Yes, this is agreed with Ann, we just focus on those for now on.

File: src/test/java/uk/gov/pay/connector/it/CardDetailsResourceITest.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    public void returnErrorAndDoNotUpdateChargeStatusIfSomeCardDetailsHaveAlreadyBeenSubmitted() throws Exception {
+
+        String chargeId = "12345669385794877";
+        setupCharge(chargeId);
+
+        given().port(app.getLocalPort())
+                .contentType(JSON)
+                .body(validCardDetails)
+                .post(cardUrlForChargeId(chargeId))
+                .then()
```
Comment (mikemey) [inline src/test/java/uk/gov/pay/connector/it/CardDetailsResourceITest.java]: Please avoid code duplication here - all the `given() .... .then()` blocks are very similar.

File: src/test/java/uk/gov/pay/connector/it/CardDetailsResourceITest.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+        String chargeId = "8172643872964549";
+        setupCharge(chargeId);
+
+        StringBuilder cardWithInvalidExpiryDateFormat = new StringBuilder();
+        cardWithInvalidExpiryDateFormat.append("{");
+        cardWithInvalidExpiryDateFormat.append("\"card_number\":\"4242424242424242\",");
+        cardWithInvalidExpiryDateFormat.append("\"cvc\":\"123\",");
+        cardWithInvalidExpiryDateFormat.append("\"expiry_date\":\"1299\"");
+        cardWithInvalidExpiryDateFormat.append("}");
```
Comment (mikemey) [inline src/test/java/uk/gov/pay/connector/it/CardDetailsResourceITest.java]: Please avoid code duplication - here with method `validCardDetails`

File: src/main/java/uk/gov/pay/connector/model/ChargeStatus.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package uk.gov.pay.connector.model;
+
+public enum ChargeStatus {
+    CREATED("CREATED"),
+    AUTHORIZATION_SUBMITTED("AUTHORIZATION SUBMITTED"),
```
Comment (davidillsley) [inline src/main/java/uk/gov/pay/connector/model/ChargeStatus.java:5]: Can we not use strings which are shouting and with spaces?

**Comment by SiliconMeeple** [comment]
rrun

File: src/test/java/uk/gov/pay/connector/it/CardDetailsResourceITest.java (hunk 5 above)
Comment (chaphilippe) [inline src/test/java/uk/gov/pay/connector/it/CardDetailsResourceITest.java]: done

File: src/test/java/uk/gov/pay/connector/it/CardDetailsResourceITest.java (hunk 4 above)
Comment (chaphilippe) [inline src/test/java/uk/gov/pay/connector/it/CardDetailsResourceITest.java]: Tests have been updated

File: src/main/java/uk/gov/pay/connector/resources/CardDetailsResource.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+    }
+
+    private boolean hasStatusCreated(Map<String, Object> charge) {
+        return ChargeStatus.CREATED.getValue().equals(charge.get("status"));
+    }
+
+    private Response responseWithInvalidCardDetails() {
+        return ResponseUtil.badResponse(String.format("Values do not match expected format/length."));
```
Comment (mikemey) [inline src/main/java/uk/gov/pay/connector/resources/CardDetailsResource.java]: String.format without parameters -> please remove.

**Comment by mikemey** [comment]
:+1:

**Pull request merged**